Panic due to mismatched footer size with Go 1.27

Upstream Go CL https://go.dev/cl/707355 (to be released in Go 1.27) optimizes the compress/flate package. After this change, compression output is no longer necessarily byte-for-byte identical to Go 1.26, so callers expecting specific output may fail. 

The estargz package appears to have such a dependency. For example, [`gzipFooterBytes`](https://github.com/containerd/stargz-snapshotter/blob/main/estargz/gzip.go#L102) expects compression of a specific payload to be exactly 51 bytes. (Note that compress/gzip uses compress/flate internally.)

There is no Go 1.27 RC for a few weeks yet. The easiest way to test with 1.27 is probably to use [golang.org/dl/gotip](https://pkg.go.dev/golang.org/dl/gotip):

```
$ go install golang.org/dl/gotip@latest
$ gotip download
$ ?   	github.com/containerd/stargz-snapshotter/analyzer	[no test files]
?   	github.com/containerd/stargz-snapshotter/analyzer/fanotify	[no test files]
?   	github.com/containerd/stargz-snapshotter/analyzer/fanotify/conn	[no test files]
?   	github.com/containerd/stargz-snapshotter/analyzer/fanotify/service	[no test files]
ok  	github.com/containerd/stargz-snapshotter/analyzer/recorder	0.771s
ok  	github.com/containerd/stargz-snapshotter/cache	0.025s
ok  	github.com/containerd/stargz-snapshotter/fs	0.061s
?   	github.com/containerd/stargz-snapshotter/fs/config	[no test files]
--- FAIL: TestLayer (0.01s)
    --- FAIL: TestLayer/testPrefetch-no_prefetch-externaltoc-gzip-bestspeed-default (0.01s)
panic: footer buffer = 43, not 46 [recovered, repanicked]

goroutine 11 [running]:
testing.tRunner.func1.2({0x10d7010, 0x15585919c010})
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2128 +0x232
testing.tRunner.func1()
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2131 +0x349
panic({0x10d7010?, 0x15585919c010?})
	/usr/local/google/home/mpratt/sdk/gotip/src/runtime/panic.go:859 +0x125
github.com/containerd/stargz-snapshotter/estargz/externaltoc.gzipFooterBytes()
	/tmp/stargz-snapshotter/estargz/externaltoc/externaltoc.go:143 +0x1ee
github.com/containerd/stargz-snapshotter/estargz/externaltoc.(*GzipCompressor).WriteTOCAndFooter(0x15585919d020, {0x11766a0, 0x1558593a80c0}, 0x20?, 0x453d85?, {0x0?, 0x0?})
	/tmp/stargz-snapshotter/estargz/externaltoc/externaltoc.go:103 +0x24d
github.com/containerd/stargz-snapshotter/estargz.tocAndFooter({0x7f16b00cf9b8, 0x15585919d040}, 0x1558593a60e0, 0x13d)
	/tmp/stargz-snapshotter/estargz/build.go:367 +0x6b
github.com/containerd/stargz-snapshotter/estargz.closeWithCombine({0x15585919d0b0, 0x2, 0x0?})
	/tmp/stargz-snapshotter/estargz/build.go:362 +0x527
github.com/containerd/stargz-snapshotter/estargz.Build(0x1558592c01e0, {0x155859294460, 0x4, 0x1558592961e0?})
	/tmp/stargz-snapshotter/estargz/build.go:271 +0xa0d
github.com/containerd/stargz-snapshotter/util/testutil.BuildEStargz({0x15585919cfe0, 0x1, 0x1}, {0x1558590e0e18, 0x1, 0x0?})
	/tmp/stargz-snapshotter/util/testutil/estargz.go:60 +0x258
github.com/containerd/stargz-snapshotter/fs/layer.testPrefetch.func2(0x15585906a930)
	/tmp/stargz-snapshotter/fs/layer/testutil.go:256 +0x2cf
github.com/containerd/stargz-snapshotter/fs/layer.(*TestRunner).Run.func1({0x117bcf0, 0x155859146fc8})
	/tmp/stargz-snapshotter/fs/layer/testutil.go:100 +0x99
github.com/containerd/stargz-snapshotter/fs/layer.TestLayer.func1.1(0x155859146fc8?)
	/tmp/stargz-snapshotter/fs/layer/layer_test.go:43 +0x22
testing.tRunner(0x155859146fc8, 0x15585919d050)
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2190 +0xea
created by testing.(*T).Run in goroutine 10
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2255 +0x4d4
FAIL	github.com/containerd/stargz-snapshotter/fs/layer	0.088s
?   	github.com/containerd/stargz-snapshotter/fs/metrics/common	[no test files]
?   	github.com/containerd/stargz-snapshotter/fs/metrics/layer	[no test files]
--- FAIL: TestReader (0.05s)
    --- FAIL: TestReader/reading_single_chunk_at_top_of_2nd_chunk_in_1_chunk_file_with_clean_cache_gzip-bestspeed (0.01s)
panic: footer buffer = 48, not 51 [recovered, repanicked]

goroutine 76 [running]:
testing.tRunner.func1.2({0xdcf2d8, 0x15c367a08300})
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2128 +0x232
testing.tRunner.func1()
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2131 +0x349
panic({0xdcf2d8?, 0x15c367a08300?})
	/usr/local/google/home/mpratt/sdk/gotip/src/runtime/panic.go:859 +0x125
github.com/containerd/stargz-snapshotter/estargz.gzipFooterBytes(0xef)
	/tmp/stargz-snapshotter/estargz/gzip.go:115 +0x279
github.com/containerd/stargz-snapshotter/estargz.(*GzipCompressor).WriteTOCAndFooter(0x15c367a7c3f0, {0xe3e9d8, 0x15c368094390}, 0xef, 0x7fdb62acb950?, {0x0, 0x0})
	/tmp/stargz-snapshotter/estargz/gzip.go:95 +0x2c8
github.com/containerd/stargz-snapshotter/estargz.tocAndFooter({0x7fdb17aeea00, 0x15c367a08180}, 0x15c367aa4640, 0xef)
	/tmp/stargz-snapshotter/estargz/build.go:367 +0x6b
github.com/containerd/stargz-snapshotter/estargz.closeWithCombine({0x15c3680204f8, 0x3, 0x0?})
	/tmp/stargz-snapshotter/estargz/build.go:362 +0x527
github.com/containerd/stargz-snapshotter/estargz.Build(0x15c367a71e00, {0x15c367a08230, 0x2, 0x15c368086120?})
	/tmp/stargz-snapshotter/estargz/build.go:271 +0xa0d
github.com/containerd/stargz-snapshotter/util/testutil.BuildEStargz({0x15c367a08220, 0x1, 0x1}, {0x15c367900c30, 0x1, 0x38?})
	/tmp/stargz-snapshotter/util/testutil/estargz.go:60 +0x258
github.com/containerd/stargz-snapshotter/fs/reader.makeFile({0xe42c68, 0x15c3680203f0}, {0x8fc5c6?, 0x3?, 0xa?}, 0x3, 0xe456c0, {0xe43ae8, 0x15c367a08180})
	/tmp/stargz-snapshotter/fs/reader/testutil.go:258 +0x265
github.com/containerd/stargz-snapshotter/fs/reader.testFileReadAt.func1(0x15c3680203f0)
	/tmp/stargz-snapshotter/fs/reader/testutil.go:164 +0x5af
github.com/containerd/stargz-snapshotter/fs/reader.(*TestRunner).Run.func1({0xe42c28, 0x15c3679e6d88})
	/tmp/stargz-snapshotter/fs/reader/testutil.go:91 +0x99
github.com/containerd/stargz-snapshotter/fs/reader.TestReader.func1.1(0x15c3679e6d88?)
	/tmp/stargz-snapshotter/fs/reader/reader_test.go:42 +0x22
testing.tRunner(0x15c3679e6d88, 0x15c367a081f0)
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2190 +0xea
created by testing.(*T).Run in goroutine 10
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2255 +0x4d4
FAIL	github.com/containerd/stargz-snapshotter/fs/reader	0.104s
ok  	github.com/containerd/stargz-snapshotter/fs/remote	11.704s
?   	github.com/containerd/stargz-snapshotter/fs/source	[no test files]
ok  	github.com/containerd/stargz-snapshotter/fusemanager	0.145s
?   	github.com/containerd/stargz-snapshotter/fusemanager/api	[no test files]
?   	github.com/containerd/stargz-snapshotter/metadata	[no test files]
--- FAIL: TestReader (0.00s)
    --- FAIL: TestReader/empty-gzip-no-compression (0.00s)
panic: footer buffer = 48, not 51 [recovered, repanicked]

goroutine 9 [running]:
testing.tRunner.func1.2({0x9d63a0, 0xa8feab32050})
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2128 +0x232
testing.tRunner.func1()
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2131 +0x349
panic({0x9d63a0?, 0xa8feab32050?})
	/usr/local/google/home/mpratt/sdk/gotip/src/runtime/panic.go:859 +0x125
github.com/containerd/stargz-snapshotter/estargz.gzipFooterBytes(0x437)
	/tmp/stargz-snapshotter/estargz/gzip.go:115 +0x279
github.com/containerd/stargz-snapshotter/estargz.(*GzipCompressor).WriteTOCAndFooter(0xa8feaa3bbe8, {0xa16230, 0xa8feab261b0}, 0x437, 0x7f6f2b831148?, {0x0, 0x0})
	/tmp/stargz-snapshotter/estargz/gzip.go:95 +0x2c8
github.com/containerd/stargz-snapshotter/estargz.tocAndFooter({0x7f6f2b835088, 0xa8feaa5c860}, 0xa8feab24080, 0x437)
	/tmp/stargz-snapshotter/estargz/build.go:367 +0x6b
github.com/containerd/stargz-snapshotter/estargz.closeWithCombine({0xa8feaa5c8c0, 0x2, 0x0?})
	/tmp/stargz-snapshotter/estargz/build.go:362 +0x527
github.com/containerd/stargz-snapshotter/estargz.Build(0xa8feaae1260, {0xa8fea8f41e8, 0x1, 0xa8fea97c300?})
	/tmp/stargz-snapshotter/estargz/build.go:271 +0xa0d
github.com/containerd/stargz-snapshotter/util/testutil.BuildEStargz({0xa984a0, 0x0, 0x0}, {0xa8fea959e58, 0x2, 0x0?})
	/tmp/stargz-snapshotter/util/testutil/estargz.go:60 +0x258
github.com/containerd/stargz-snapshotter/metadata/testutil.TestReader.func1(0xa8fea904390)
	/tmp/stargz-snapshotter/metadata/testutil/testutil.go:330 +0x665
github.com/containerd/stargz-snapshotter/metadata/testutil.(*TestRunner).Run.func1({0xa17970, 0xa8feaa12488})
	/tmp/stargz-snapshotter/metadata/testutil/testutil.go:84 +0x99
github.com/containerd/stargz-snapshotter/metadata/memory.TestReader.func1.1(0xa8feaa12488?)
	/tmp/stargz-snapshotter/metadata/memory/reader_test.go:38 +0x22
testing.tRunner(0xa8feaa12488, 0xa8feaa5c870)
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2190 +0xea
created by testing.(*T).Run in goroutine 8
	/usr/local/google/home/mpratt/sdk/gotip/src/testing/testing.go:2255 +0x4d4
FAIL	github.com/containerd/stargz-snapshotter/metadata/memory	0.072s
?   	github.com/containerd/stargz-snapshotter/metadata/testutil	[no test files]
?   	github.com/containerd/stargz-snapshotter/nativeconverter	[no test files]
time="2026-05-22T16:39:40-04:00" level=warning msg="failed to set containerd.io/uncompressed label for sha256:0e03bdcc26d7a9a57ef3b6f1bf1a210cff6239bff7c8cac72435984032851689" error="update not supported on immutable content store: failed precondition"
panic: footer buffer = 48, not 51

goroutine 81 [running]:
github.com/containerd/stargz-snapshotter/estargz.gzipFooterBytes(0xa72)
	/tmp/stargz-snapshotter/estargz/gzip.go:115 +0x279
github.com/containerd/stargz-snapshotter/estargz.(*GzipCompressor).WriteTOCAndFooter(0x1f0b2b561610, {0xc06ba0, 0x1f0b2b48a660}, 0xa72, 0x44ea45?, {0x0, 0x0})
	/tmp/stargz-snapshotter/estargz/gzip.go:95 +0x2c8
github.com/containerd/stargz-snapshotter/estargz.tocAndFooter({0x7f73922190b8, 0x1f0b2b480d00}, 0x1f0b2b4c8000, 0xa72)
	/tmp/stargz-snapshotter/estargz/build.go:367 +0x6b
github.com/containerd/stargz-snapshotter/estargz.closeWithCombine({0x1f0b2b459110, 0x3, 0x1f0b2b480ae0?})
	/tmp/stargz-snapshotter/estargz/build.go:362 +0x527
github.com/containerd/stargz-snapshotter/estargz.Build(0x1f0b2b703c50, {0x1f0b2b480cf0, 0x2, 0x9e1?})
	/tmp/stargz-snapshotter/estargz/build.go:271 +0xa0d
github.com/containerd/stargz-snapshotter/nativeconverter/estargz.LayerConvertFunc.func1({0xc09a78, 0x1f0b2b524550}, {0xc0aff8, 0x1f0b2b684980}, {{0x1f0b2b7b4880, 0x31}, {0x1f0b2b47b900, 0x47}, 0x9e1, {0x0, ...}, ...})
	/tmp/stargz-snapshotter/nativeconverter/estargz/estargz.go:83 +0x412
github.com/containerd/containerd/v2/core/images/converter.(*defaultConverter).convertLayer(0x1f0b2b5ea400?, {0xc09a78?, 0x1f0b2b524550?}, {0xc0aff8?, 0x1f0b2b684980?}, {{0x1f0b2b7b4880, 0x31}, {0x1f0b2b47b900, 0x47}, 0x9e1, ...})
	/usr/local/google/home/mpratt/go/pkg/mod/github.com/containerd/containerd/v2@v2.2.3/core/images/converter/default.go:140 +0xb0
github.com/containerd/containerd/v2/core/images/converter.(*defaultConverter).convert(0x1f0b2b524230, {0xc09a78, 0x1f0b2b524550}, {0xc0aff8, 0x1f0b2b684980}, {{0x1f0b2b7b4880, 0x31}, {0x1f0b2b47b900, 0x47}, 0x9e1, ...})
	/usr/local/google/home/mpratt/go/pkg/mod/github.com/containerd/containerd/v2@v2.2.3/core/images/converter/default.go:91 +0x3b6
github.com/containerd/containerd/v2/core/images/converter.(*defaultConverter).convertManifest.func1()
	/usr/local/google/home/mpratt/go/pkg/mod/github.com/containerd/containerd/v2@v2.2.3/core/images/converter/default.go:173 +0x132
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/usr/local/google/home/mpratt/go/pkg/mod/golang.org/x/sync@v0.20.0/errgroup/errgroup.go:93 +0x50
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 67
	/usr/local/google/home/mpratt/go/pkg/mod/golang.org/x/sync@v0.20.0/errgroup/errgroup.go:78 +0xa5
FAIL	github.com/containerd/stargz-snapshotter/nativeconverter/estargz	0.589s
time="2026-05-22T16:39:40-04:00" level=warning msg="failed to set containerd.io/uncompressed label for sha256:0e03bdcc26d7a9a57ef3b6f1bf1a210cff6239bff7c8cac72435984032851689" error="update not supported on immutable content store: failed precondition"
panic: footer buffer = 43, not 46

goroutine 35 [running]:
github.com/containerd/stargz-snapshotter/estargz/externaltoc.gzipFooterBytes()
	/tmp/stargz-snapshotter/estargz/externaltoc/externaltoc.go:143 +0x1ee
github.com/containerd/stargz-snapshotter/estargz/externaltoc.(*GzipCompressor).WriteTOCAndFooter(0xf0813dae270, {0x109c678, 0xf0813eb41b0}, 0x8?, 0x453105?, {0x0?, 0x0?})
	/tmp/stargz-snapshotter/estargz/externaltoc/externaltoc.go:103 +0x24d
github.com/containerd/stargz-snapshotter/estargz.tocAndFooter({0x7fe234263d00, 0xf0813dae280}, 0xf0813eb6160, 0xb24)
	/tmp/stargz-snapshotter/estargz/build.go:367 +0x6b
github.com/containerd/stargz-snapshotter/estargz.closeWithCombine({0xf0813dae2d0, 0x2, 0x0?})
	/tmp/stargz-snapshotter/estargz/build.go:362 +0x527
github.com/containerd/stargz-snapshotter/estargz.Build(0xf0813daa210, {0xf0813dae2a0, 0x2, 0x47?})
	/tmp/stargz-snapshotter/estargz/build.go:271 +0xa0d
github.com/containerd/stargz-snapshotter/nativeconverter/estargz.LayerConvertFunc.func1({0x10a1638, 0xf0813d42550}, {0x10a3868, 0xf0813ed21c0}, {{0xf0813ee6a40, 0x31}, {0xf0813ec5450, 0x47}, 0x9e1, {0x0, ...}, ...})
	/tmp/stargz-snapshotter/nativeconverter/estargz/estargz.go:83 +0x4e9
github.com/containerd/stargz-snapshotter/nativeconverter/estargz/externaltoc.layerConvert.func1({0x10a1638, 0xf0813d42550}, {0x10a3868, 0xf0813ed21c0}, {{0xf0813ee6a40, 0x31}, {0xf0813ec5450, 0x47}, 0x9e1, {0x0, ...}, ...})
	/tmp/stargz-snapshotter/nativeconverter/estargz/externaltoc/converter.go:121 +0x10a
github.com/containerd/containerd/v2/core/images/converter.(*defaultConverter).convertLayer(0x43c9f7?, {0x10a1638?, 0xf0813d42550?}, {0x10a3868?, 0xf0813ed21c0?}, {{0xf0813ee6a40, 0x31}, {0xf0813ec5450, 0x47}, 0x9e1, ...})
	/usr/local/google/home/mpratt/go/pkg/mod/github.com/containerd/containerd/v2@v2.2.3/core/images/converter/default.go:140 +0xb0
github.com/containerd/containerd/v2/core/images/converter.(*defaultConverter).convert(0xf0813d42230, {0x10a1638, 0xf0813d42550}, {0x10a3868, 0xf0813ed21c0}, {{0xf0813ee6a40, 0x31}, {0xf0813ec5450, 0x47}, 0x9e1, ...})
	/usr/local/google/home/mpratt/go/pkg/mod/github.com/containerd/containerd/v2@v2.2.3/core/images/converter/default.go:91 +0x3b6
github.com/containerd/containerd/v2/core/images/converter.(*defaultConverter).convertManifest.func1()
	/usr/local/google/home/mpratt/go/pkg/mod/github.com/containerd/containerd/v2@v2.2.3/core/images/converter/default.go:173 +0x132
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/usr/local/google/home/mpratt/go/pkg/mod/golang.org/x/sync@v0.20.0/errgroup/errgroup.go:93 +0x50
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 84
	/usr/local/google/home/mpratt/go/pkg/mod/golang.org/x/sync@v0.20.0/errgroup/errgroup.go:78 +0xa5
FAIL	github.com/containerd/stargz-snapshotter/nativeconverter/estargz/externaltoc	0.533s
ok  	github.com/containerd/stargz-snapshotter/nativeconverter/zstdchunked	0.462s
?   	github.com/containerd/stargz-snapshotter/recorder	[no test files]
?   	github.com/containerd/stargz-snapshotter/script/benchmark/hello-bench/src/go	[no test files]
?   	github.com/containerd/stargz-snapshotter/service	[no test files]
?   	github.com/containerd/stargz-snapshotter/service/keychain/cri	[no test files]
?   	github.com/containerd/stargz-snapshotter/service/keychain/dockerconfig	[no test files]
?   	github.com/containerd/stargz-snapshotter/service/keychain/keychainconfig	[no test files]
?   	github.com/containerd/stargz-snapshotter/service/keychain/kubeconfig	[no test files]
?   	github.com/containerd/stargz-snapshotter/service/plugin	[no test files]
?   	github.com/containerd/stargz-snapshotter/service/plugincore	[no test files]
?   	github.com/containerd/stargz-snapshotter/service/resolver	[no test files]
ok  	github.com/containerd/stargz-snapshotter/snapshot	0.055s
?   	github.com/containerd/stargz-snapshotter/store	[no test files]
?   	github.com/containerd/stargz-snapshotter/store/pb	[no test files]
ok  	github.com/containerd/stargz-snapshotter/task	0.914s
ok  	github.com/containerd/stargz-snapshotter/util/cacheutil	9.023s
?   	github.com/containerd/stargz-snapshotter/util/containerdutil	[no test files]
ok  	github.com/containerd/stargz-snapshotter/util/decompressutil	0.031s
?   	github.com/containerd/stargz-snapshotter/util/ioutils	[no test files]
?   	github.com/containerd/stargz-snapshotter/util/namedmutex	[no test files]
?   	github.com/containerd/stargz-snapshotter/util/testutil	[no test files]
?   	github.com/containerd/stargz-snapshotter/version	[no test files]
FAIL
```

I am not familiar with this format at all, but given that you are writing such a small payload you may be better off manually writing out the entire payload without involving compress/gzip.

--- a/estargz/externaltoc/externaltoc.go
+++ b/estargz/externaltoc/externaltoc.go
@@ -126,23 +126,20 @@ const FooterSize = 46
 
 // gzipFooterBytes returns the 104 bytes footer.
 func gzipFooterBytes() ([]byte, error) {
-	buf := bytes.NewBuffer(make([]byte, 0, FooterSize))
-	gz, _ := gzip.NewWriterLevel(buf, gzip.NoCompression) // MUST be NoCompression to keep 51 bytes
-
-	// Extra header indicating the offset of TOCJSON
+	// Extra header indicating the external toc
 	// https://tools.ietf.org/html/rfc1952#section-2.3.1.1
 	header := make([]byte, 4)
 	header[0], header[1] = 'S', 'G'
 	subfield := "STARGZEXTERNALTOC"                                   // len("STARGZEXTERNALTOC") = 17
 	binary.LittleEndian.PutUint16(header[2:4], uint16(len(subfield))) // little-endian per RFC1952
-	gz.Extra = append(header, []byte(subfield)...)
-	if err := gz.Close(); err != nil {
-		return nil, err
+	extra := append(header, []byte(subfield)...)
+
+	buf := estargz.CreateGzipFooter(extra)
+
+	if len(buf) != FooterSize {
+		panic(fmt.Sprintf("footer buffer = %d, not %d", len(buf), FooterSize))
 	}
-	if buf.Len() != FooterSize {
-		panic(fmt.Sprintf("footer buffer = %d, not %d", buf.Len(), FooterSize))
-	}
-	return buf.Bytes(), nil
+	return buf, nil
 }
 
 func NewGzipDecompressor(provideTOCFunc func() ([]byte, error)) *GzipDecompressor {

--- a/estargz/gzip.go
+++ b/estargz/gzip.go
@@ -100,21 +100,52 @@ func (gc *GzipCompressor) WriteTOCAndFooter(w io.Writer, off int64, toc *JTOC, d
 
 // gzipFooterBytes returns the 51 bytes footer.
 func gzipFooterBytes(tocOff int64) []byte {
-	buf := bytes.NewBuffer(make([]byte, 0, FooterSize))
-	gz, _ := gzip.NewWriterLevel(buf, gzip.NoCompression) // MUST be NoCompression to keep 51 bytes
-
 	// Extra header indicating the offset of TOCJSON
 	// https://tools.ietf.org/html/rfc1952#section-2.3.1.1
 	header := make([]byte, 4)
 	header[0], header[1] = 'S', 'G'
 	subfield := fmt.Sprintf("%016xSTARGZ", tocOff)
 	binary.LittleEndian.PutUint16(header[2:4], uint16(len(subfield))) // little-endian per RFC1952
-	gz.Extra = append(header, []byte(subfield)...)
-	gz.Close()
-	if buf.Len() != FooterSize {
-		panic(fmt.Sprintf("footer buffer = %d, not %d", buf.Len(), FooterSize))
+	extra := append(header, []byte(subfield)...)
+
+	buf := CreateGzipFooter(extra)
+	if len(buf) != FooterSize {
+		panic(fmt.Sprintf("footer buffer = %d, not %d", len(buf), FooterSize))
 	}
-	return buf.Bytes()
+	return buf
+}
+
+// CreateGzipFooter creates eStargz's footer with the specified extra field.
+func CreateGzipFooter(extra []byte) []byte {
+	// Gzip header (10 bytes)
+	// https://datatracker.ietf.org/doc/html/rfc1952
+	buf := make([]byte, 10)
+	buf[0] = 0x1f // ID1
+	buf[1] = 0x8b // ID2
+	buf[2] = 8    // "deflate"
+	buf[3] = 0x4  // FEXTRA
+	// MTIME = 0
+	// XFL = 0
+	buf[9] = 255 // unknown
+
+	// Extra field
+	xlen := make([]byte, 2)
+	binary.LittleEndian.PutUint16(xlen, uint16(len(extra)))
+	buf = append(buf, append(xlen, extra...)...)
+
+	// Flate header (5 bytes)
+	// https://datatracker.ietf.org/doc/html/rfc1951
+	flateHeader := []byte{
+		1,    // BFINAL = 1 (last block), BFTYPE = 0 (no compression), remaining bits are ignored
+		0, 0, // LEN = 0
+		0xff, 0xff, // NLEN (the one's complement of LEN)
+	}
+	buf = append(buf, flateHeader...)
+
+	// Gzip footer (8 bytes): CRC32 = 0, ISIZE = 0
+	buf = append(buf, make([]byte, 8)...)
+
+	return buf
 }
 
 type GzipDecompressor struct{}

--- a/estargz/gzip_test.go
+++ b/estargz/gzip_test.go
@@ -151,12 +151,9 @@ func TestGzipParseFooterInvalidExtra(t *testing.T) {
 }
 
 func legacyFooterBytes(tocOff int64) []byte {
-	buf := bytes.NewBuffer(make([]byte, 0, legacyFooterSize))
-	gz, _ := gzip.NewWriterLevel(buf, gzip.NoCompression)
-	gz.Extra = fmt.Appendf(nil, "%016xSTARGZ", tocOff)
-	gz.Close()
-	if buf.Len() != legacyFooterSize {
-		panic(fmt.Sprintf("footer buffer = %d, not %d", buf.Len(), legacyFooterSize))
+	buf := CreateGzipFooter(fmt.Appendf(nil, "%016xSTARGZ", tocOff))
+	if len(buf) != legacyFooterSize {
+		panic(fmt.Sprintf("footer buffer = %d, not %d", len(buf), legacyFooterSize))
 	}
-	return buf.Bytes()
+	return buf
 }

